app/controllers/mixins/breadcrumbs_mixin.rb

```diff
@@ -33,33 +34,58 @@ def data_for_breadcrumbs
           breadcrumbs.push(:title => @title)
         end
       else
+        self.x_node_text = params[:text] if action_name == "tree_select" # get text of the active node
+
         # Append breadcrumb from title of the accordion (eg "Policies")
-        breadcrumbs.push(:title => accord_title, :key => "#{accord_name}_accord", :action => "accordion_select") if accord_title
+        breadcrumbs.push(:title => accord_title, :key => "root") if accord_title
+
+        # On special page (tagitems, politems, etc.)
+        if @tagitems || @politems || @ownershipitems || @retireitems
+          breadcrumbs.push(special_page_breadcrumb(@tagitems || @politems || @ownershipitems || @retireitems, true)) unless options[:hide_special_item]
+          breadcrumbs.push(:title => @right_cell_text)
+        else
+          # Ancestry parents breadcrumbs (only in services)
+          breadcrumbs.concat(ancestry_parents(options[:ancestry], options[:record_info], options[:record_title])) if options[:ancestry]
 
-        # Append breadcrumbs created from the tree (eg "All policies > Red Hat policies > Policy 1")
-        breadcrumbs_from_tree = build_breadcrumbs_from_tree
-        breadcrumbs.concat(breadcrumbs_from_tree)
+          # Priority 1. Include record
+          #          2. TreeNode.text
+          #          3. Item from tree (if it's built)
+          #          4. Fallbacks (sent text, title of the record, title_for_breadcrumbs)
+          #
+          # Overrides: if x_root show then show title_for_breadcrumbs
 
-        if (options[:include_record] || breadcrumbs_from_tree.blank?) && options[:record_info].present? && options[:record_info][options[:record_title]] &&
-           !(@tagitems || @politems || @ownershipitems || @retireitems)
-          # Append breadcrumb created from record_info if included or if result from tree was empty (eg filters page)
+          # Include items (for trees, which does not have final nodes: VMs etc.)
+          if options[:include_record] && options[:record_info].present?
+            title = options[:record_info][options[:record_title]]
+            key = TreeBuilder.build_node_id(options[:record_info])
+          end
```

Comment:
I don't see why this shouldn't be used with `TreeNode.new()` and then calling `text` and `key` on it.

Reply:
Maybe you have something special for the title, idk ... please look into this more.